src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs

```diff
@@ -0,0 +1,176 @@
+
+                currBlockNumber--;
+            }
+
+            return gasPrices;
+        }
+        
+        private SortedSet<UInt256> eth_initializeValues(Block headBlock, Block genesisBlock, out long blocksToGoBack, out int percentile,
+            out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest)
+        {
+            Comparer<UInt256> comparer = Comparer<UInt256>.Create(((a, b) =>
+            {
+                int res = a.CompareTo(b);
+                return res == 0 ? 1 : res;
+            }));
+            SortedSet<UInt256> gasPrices = new(comparer); //allows duplicates
+            blocksToGoBack = 5;
+            percentile = 20;
+            threshold = blocksToGoBack * 2;
+            currBlockNumber = headBlock!.Number;
+            genBlockNumber = genesisBlock!.Number;
+            gasPriceLatest = eth_latestGasPrice(headBlock!.Number, genesisBlock!.Number);
+            return gasPrices;
+        }
+        
+        private static bool eth_initalChecks(Block? headBlock, Block? genesisBlock, out ResultWrapper<UInt256?> resultWrapper)
+        {
+            if (headBlock == null || genesisBlock == null)
+            {
+                {
+                    resultWrapper = ResultWrapper<UInt256?>.Fail("The head block or genesis block had a null value.");
+                    return true;
+                }
+            }
+            
+            if (_lastPrice != null && _lastHeadBlock != null)
+            {
+                if (headBlock.Hash == _lastHeadBlock.Hash)
+                {
+                    {
+                        resultWrapper = ResultWrapper<UInt256?>.Success(_lastPrice);
+                        #if DEBUG
+                            resultWrapper.ErrorCode = -1;
+                        #endif
+                        return true;
+                    }
+                }
+            }
+
+            resultWrapper = ResultWrapper<UInt256?>.Success(UInt256.Zero);
+            return false;
+        }
+        
+        public ResultWrapper<UInt256?> eth_gasPrice(UInt256? ignoreUnder = null)
+        {
+            Block? headBlock = _blockFinder.FindHeadBlock();
+            Block? genesisBlock = _blockFinder.FindGenesisBlock();
+            if (eth_initalChecks(headBlock, genesisBlock, out ResultWrapper<UInt256?> resultWrapper))
+            {
+                return resultWrapper;
+            }
+
+            SortedSet<UInt256> gasPrices = eth_initializeValues(headBlock, genesisBlock, out long blocksToGoBack,
+                out int percentile, out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest);
+            if (gasPriceLatest == null)
+            {
+                return ResultWrapper<UInt256?>.Fail("gasPriceLatest was not set properly.");
+            }
+            
+            gasPrices = eth_AddingPreviousBlockTx(ignoreUnder, blocksToGoBack, currBlockNumber, genBlockNumber, 
```

Comment:
it is important to take into consideration not only gas prices but also gas used for the previous block and when calculating percentile, take the last 20% of block space and not the last 20% of gas prices

Reply:
Is this a change for EIP-1559 and Is this something we want to do differently than Geth?

If we do want to do it differently, should I create an array of block space (capacity - number of transactions in block) in addition to the gas used by the previous block to calculate the new gas price? 

If we call the gas used by the previous block _g_, the 20th percentile of sorted block space _s_, and the number of tx in the previous block _n_, then_ could the new gas be calculated as: (_g_ / _n_ ) * (_s_ / (capacity / 2))? 

That formula would allow us to increase the average gas price from the previous block if the number of tx is greater than half the capacity, and reduce it if it is less than half the capacity.